### ISSUE TITLE
`O(n)` worst case sampling runtime

### ISSUE THREAD

**Issue by LilithHafner** [Tortar/DynamicSampling.jl#21]
The worst case sample time is also `O(n)` (or `O(1)` with a large constant factor, if you prefer to think about it that way), achieved when the list of levels starts with a lot of zeros:

```julia
julia> using DynamicSampling, Chairmarks

julia> ds = DynamicSampler()
DynamicSampler(Tuple{Int64, Float64}[])

julia> push!(ds, 1, 1.0)
DynamicSampler([(1, 1.0)])

julia> @b ds rand
8.429 ns

julia> for i in 2:1000
           push!(ds, i, 2.0^i)
       end

julia> @b ds rand
16.424 ns

julia> for i in 2:1000
           delete!(ds, i)
       end

julia> @b ds rand
1.292 μs (1 allocs: 48 bytes)
```

It's possible that this could be fixed by performing a partial sort while traversing the list during sampling, though maybe there's another way with less sampling overhead.

**Comment by LilithHafner** [comment]
IIUC, the best this algorithm can get is `O(log(n))` sampling time because of the case where weights are of the form `[1, 1/2, 1/2, 1/4, 1/4, 1/4, 1/4, ...]` and level selection alone takes `log(n)` time.

**Comment by Tortar** [comment]
> IIUC, the best this algorithm can get is `O(log(n))` sampling time because of the case where weights are of the form `[1, 1/2, 1/2, 1/4, 1/4, 1/4, 1/4, ...]` and level selection alone takes `log(n)` time.

Yes my earlier comment was wrong, you are right, it actually requires `O(log(n))` as you say, and we can't do better than that for this case
